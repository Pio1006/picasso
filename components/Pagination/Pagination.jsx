import React from 'react'
import PropTypes from 'prop-types'

import Button from '../Button'
import { getRange, ELLIPSIS, FIRST_PAGE, ONE_PAGE } from './range-utils'

const NAVIGATION = {
  FIRST: 'FIRST',
  LAST: 'LAST',
  PREVIOUS: 'PREVIOUS',
  NEXT: 'NEXT'
}
const SIBLING_COUNT = 3

class Pagination extends React.PureComponent {
  static propTypes = {
    activePage: PropTypes.number.isRequired,
    disabled: PropTypes.bool,
    onPageChange: PropTypes.func.isRequired,
    totalPages: PropTypes.number.isRequired
  }

  static defaultProps = {
    disabled: false
  }

  handleChange (navigation) {
    const { onPageChange, totalPages, activePage } = this.props
    let page

    switch (navigation) {
      case NAVIGATION.FIRST:
        page = FIRST_PAGE
        break
      case NAVIGATION.PREVIOUS:
        page = activePage - ONE_PAGE
        break
      case NAVIGATION.NEXT:
        page = activePage + ONE_PAGE
        break
      case NAVIGATION.LAST:
        page = totalPages
        break
      default:
        page = navigation
    }

    onPageChange(page)
  }

  isFirstActive = activePage => activePage === 1
  isLastActive = (activePage, totalPages) => activePage === totalPages

  renderRange () {
    const { totalPages, activePage, disabled } = this.props

    const range = getRange({
      activePage,
      totalPages,
      siblingCount: SIBLING_COUNT
    })

    return range.map((pageItemLabel, index) => {
      return (
        <Button
          disabled={pageItemLabel === ELLIPSIS || disabled}
          key={pageItemLabel + index} // eslint-disable-line react/no-array-index-key
          onClick={() => this.handleChange(pageItemLabel)}
          variant={activePage === pageItemLabel ? 'primary' : 'basic'}
        >
          {pageItemLabel}
        </Button>
      )
    })
  }

  render () {
    const { activePage, totalPages, disabled } = this.props
    const isFirstActive = this.isFirstActive(activePage)
    const isLastActive = this.isLastActive(activePage, totalPages)

    return (
      <Button.Group>
        <Button
          disabled={isFirstActive || disabled}
          onClick={() => this.handleChange(NAVIGATION.FIRST)}
          variant='basic'
        >
          «
        </Button>
        <Button
          disabled={isFirstActive || disabled}
          onClick={() => this.handleChange(NAVIGATION.PREVIOUS)}
          variant='basic'
        >
          ⟨
        </Button>

        {this.renderRange()}

        <Button
          disabled={isLastActive || disabled}
          onClick={() => this.handleChange(NAVIGATION.NEXT)}
          variant='basic'
        >
          ⟩
        </Button>
        <Button
          disabled={isLastActive || disabled}
          onClick={() => this.handleChange(NAVIGATION.LAST)}
          variant='basic'
        >
          »
        </Button>
      </Button.Group>
    )
  }
}

export default Pagination
