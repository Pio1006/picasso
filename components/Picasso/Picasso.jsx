import { createMuiTheme, MuiThemeProvider } from '@material-ui/core/styles'
import React from 'react'
import PT from 'prop-types'

import pallete from './pallete'
import Provider from './PicassoProvider'

const picasso = {
  pallete,
  props: {
    MuiButtonBase: {
      disableRipple: true
    }
  },
  shadows: Array.apply(null, Array(25)).map(() => 'none'), // Yest, this is needed as MUI expects 25 elements in array otherwise it raises error
  typography: {
    useNextVariants: true
  }
}

const createTheme = provider => createMuiTheme(provider.theme)

const Picasso = ({ children }) => (
  <MuiThemeProvider theme={createTheme(PicassoProvider)}>
    {children}
  </MuiThemeProvider>
)

Picasso.displayName = Picasso
Picasso.propTypes = {
  children: PT.node.isRequired
}

export const PicassoProvider = new Provider(picasso)
export default Picasso
