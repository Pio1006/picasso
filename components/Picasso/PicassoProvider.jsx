export class PicassoProvider {
  constructor (theme = {}) {
    this.theme = theme
  }

  override (getOverride = () => {}) {
    const newOverride = getOverride(this.theme)

    this.extendThemeOverrides(newOverride)
  }

  withTheme (getWithTheme = () => {}) {
    return getWithTheme(this.theme)
  }

  extendThemeOverrides (newOverride) {
    const overrides = this.theme.overrides || {}

    Object.assign(overrides, newOverride)

    this.theme.overrides = overrides
  }
}

export default PicassoProvider
