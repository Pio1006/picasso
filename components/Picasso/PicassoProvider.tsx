import { Theme } from '@material-ui/core/styles/createMuiTheme'

export class PicassoProvider {
  theme: Theme

  constructor (theme: Theme) {
    this.theme = theme
  }

  override (getOverride: (theme: Theme) => any = () => ({})) {
    const newOverride = getOverride(this.theme)

    this.extendThemeOverrides(newOverride)
  }

  withTheme (getWithTheme: (theme: Theme) => any = () => ({})) {
    return getWithTheme(this.theme)
  }

  extendThemeOverrides (newOverride: Theme) {
    const overrides = this.theme.overrides || {}

    Object.assign(overrides, newOverride)

    this.theme.overrides = overrides
  }
}

export default PicassoProvider
