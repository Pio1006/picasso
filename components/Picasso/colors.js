const colors = {
  primary: {
    main: '#204ecf',
    darker: '#1542c1'
  },
  secondary: {
    main: '#00cc83'
  },
  grey: {
    50: '#dfe3e9',
    100: '#686869'
  },
  blue: {
    light: '#cad5f4'
  },
  error: {
    main: '#9f3a38',
    text: '#e0b4b4',
    background: '#fff6f6'
  }
}

export default colors
