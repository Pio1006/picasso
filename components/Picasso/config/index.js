export { default as palette } from './palette'
export { default as transitions } from './transitions'
export { default as typography } from './typography'
