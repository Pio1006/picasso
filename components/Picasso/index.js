export { PicassoProvider } from './Picasso'
export { default } from './Picasso'
