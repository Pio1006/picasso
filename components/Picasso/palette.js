const palette = {
  common: {
    black: '#000000',
    white: '#ffffff'
  },
  primary: {
    main: '#204ecf',
    dark: '#1542c1'
  },
  secondary: {
    main: '#00cc83'
  },
  text: {
    primary: '#455065'
  },
  grey: {
    50: '#dfe3e9',
    100: '#686869'
  },
  blue: {
    light: 'rgba(32,78,207,.05)'
  },
  error: {
    main: '#9f3a38',
    text: '#e0b4b4',
    background: '#fff6f6'
  }
}

export default palette
