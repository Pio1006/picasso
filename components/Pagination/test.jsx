import React from 'react'
/* eslint-disable-next-line */
import { render, fireEvent, cleanup } from 'react-testing-library'

import Pagination from './index'
import Picasso from '../Picasso'

const renderPagination = (props = {}) => {
  return render(
    <Picasso>
      <Pagination {...props} />
    </Picasso>
  )
}

afterEach(cleanup)

test('renders default', () => {
  const { container } = renderPagination({
    activePage: 5,
    totalPages: 20
  })

  expect(container).toMatchSnapshot()
})

test('renders disabled', () => {
  const { container } = renderPagination({
    activePage: 5,
    totalPages: 20,
    disabled: true
  })

  expect(container).toMatchSnapshot()
})
