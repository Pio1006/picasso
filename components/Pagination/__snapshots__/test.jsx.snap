// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders default 1`] = `
<div>
  <div
    class="ButtonGroup-root"
  >
    <button
      class="MuiButtonBase-root Button-basic Button-root"
      tabindex="0"
      type="button"
    >
      <div
        class="Button-children"
      >
        «
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root"
      tabindex="0"
      type="button"
    >
      <div
        class="Button-children"
      >
        ⟨
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root MuiButtonBase-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <div
        class="Button-children"
      >
        ...
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root"
      tabindex="0"
      type="button"
    >
      <div
        class="Button-children"
      >
        2
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root"
      tabindex="0"
      type="button"
    >
      <div
        class="Button-children"
      >
        3
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root"
      tabindex="0"
      type="button"
    >
      <div
        class="Button-children"
      >
        4
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-primary Button-root"
      tabindex="0"
      type="button"
    >
      <div
        class="Button-children"
      >
        5
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root"
      tabindex="0"
      type="button"
    >
      <div
        class="Button-children"
      >
        6
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root"
      tabindex="0"
      type="button"
    >
      <div
        class="Button-children"
      >
        7
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root"
      tabindex="0"
      type="button"
    >
      <div
        class="Button-children"
      >
        8
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root MuiButtonBase-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <div
        class="Button-children"
      >
        ...
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root"
      tabindex="0"
      type="button"
    >
      <div
        class="Button-children"
      >
        ⟩
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root"
      tabindex="0"
      type="button"
    >
      <div
        class="Button-children"
      >
        »
      </div>
    </button>
  </div>
</div>
`;

exports[`renders disabled 1`] = `
<div>
  <div
    class="ButtonGroup-root"
  >
    <button
      class="MuiButtonBase-root Button-basic Button-root MuiButtonBase-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <div
        class="Button-children"
      >
        «
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root MuiButtonBase-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <div
        class="Button-children"
      >
        ⟨
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root MuiButtonBase-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <div
        class="Button-children"
      >
        ...
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root MuiButtonBase-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <div
        class="Button-children"
      >
        2
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root MuiButtonBase-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <div
        class="Button-children"
      >
        3
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root MuiButtonBase-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <div
        class="Button-children"
      >
        4
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-primary Button-root MuiButtonBase-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <div
        class="Button-children"
      >
        5
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root MuiButtonBase-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <div
        class="Button-children"
      >
        6
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root MuiButtonBase-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <div
        class="Button-children"
      >
        7
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root MuiButtonBase-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <div
        class="Button-children"
      >
        8
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root MuiButtonBase-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <div
        class="Button-children"
      >
        ...
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root MuiButtonBase-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <div
        class="Button-children"
      >
        ⟩
      </div>
    </button>
    <button
      class="MuiButtonBase-root Button-basic Button-root MuiButtonBase-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <div
        class="Button-children"
      >
        »
      </div>
    </button>
  </div>
</div>
`;
